core/src/main/java/com/linecorp/armeria/client/DefaultEventLoopScheduler.java

```diff
@@ -0,0 +1,443 @@
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.util.ReleasableHolder;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import it.unimi.dsi.fastutil.objects.Object2IntOpenHashMap;
+
+final class DefaultEventLoopScheduler implements EventLoopScheduler {
+
+    private static final long CLEANUP_INTERVAL_NANOS = Duration.ofMinutes(1).toNanos();
+
+    private final List<EventLoop> eventLoops;
+
+    private final int maxNumEventLoops;
+
+    private final Object2IntMap<String> maxNumEventLoopsMap;
+
+    private final Map<AuthorityWithTLS, State> states = new ConcurrentHashMap<>();
+
+    private final AtomicInteger currentIndex = new AtomicInteger();
+
+    private int counter;
+    private volatile long lastCleanupTimeNanos = System.nanoTime();
+
+    DefaultEventLoopScheduler(EventLoopGroup eventLoopGroup, int maxNumEventLoops,
+                              Map<String, Integer> maxNumEventLoopsMap) {
+        eventLoops = Streams.stream(eventLoopGroup)
+                            .map(EventLoop.class::cast)
+                            .collect(toImmutableList());
+        checkArgument(maxNumEventLoops <= eventLoops.size(),
+                      "maxNumEventLoopsPerEndpoint: %s (expected: <= %s)", maxNumEventLoops,
+                      eventLoops.size());
+        if (maxNumEventLoops <= 0) {
+            this.maxNumEventLoops = eventLoops.size();
+        } else {
+            this.maxNumEventLoops = maxNumEventLoops;
+        }
+
+        for (Map.Entry<String, Integer> entry : maxNumEventLoopsMap.entrySet()) {
+            checkArgument(entry.getValue() <= eventLoops.size(),
+                          "maxNumEventLoopsPerEndpoint for %s: %s (expected: <= %s)",
+                          entry.getKey(), entry.getValue(), eventLoops.size());
+        }
+        this.maxNumEventLoopsMap = new Object2IntOpenHashMap<>(maxNumEventLoopsMap, FAST_LOAD_FACTOR);
+    }
+
+    @Override
+    public ReleasableHolder<EventLoop> acquire(Endpoint endpoint, SessionProtocol sessionProtocol) {
+        requireNonNull(endpoint, "endpoint");
+        requireNonNull(sessionProtocol, "sessionProtocol");
+        final State state = state(endpoint, sessionProtocol);
+        final Entry acquired = state.acquire();
+        cleanup();
+        return acquired;
+    }
+
+    @VisibleForTesting
+    List<Entry> entries(Endpoint endpoint, SessionProtocol sessionProtocol) {
+        return state(endpoint, sessionProtocol).entries();
+    }
+
+    private State state(Endpoint endpoint, SessionProtocol sessionProtocol) {
+        final String authority = endpoint.authority();
```

Comment:
Could we normalize the authority by appending the port number if it is same with `SessionProtocol.defaultPort()`?

```java
final String authority;
final String nextTry;
if (endpoint.isGroup()) {
    authority = endpoint.authority();
    nextTry = null;
} else {
    final Endpoint normalizedEndpoint;
    if (endpoint.hasPort()) {
        authority = endpoint.authority();
    } else if (endpoint.isIpAddrOnly() && endpoint.ipFamily() == INET6) {
        authority = '[' + endpoint.host() + "]:" + sessionProtocol.defaultPort();
    } else {
        authority = endpoint.host() + ':' + sessionProtocol.defaultPort();
    }
    nextTry = endpoint.host();
}
```

Reply:
Alternatively we could expand `AuthorityWithTLS` into a tuple of (`Endpoint.host()`, `Endpoint.port()`, TLSness) to avoid the cost of string concatenation.